### PR TITLE
[AEA-669] libp2p refactoring -  Deprecate old libp2p node

### PR DESCRIPTION
**Pull request by lrahmani**
## Proposed changes

- Add DelegateClient golang tests
- Add libp2p dht routing golang tests
- Deprecate old libp2p node
- Fix libp2p race condition (which made test_communication tests flaky)

### CODE REVIEW

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go [hunk 1]
```diff
@@ -344,11 +338,15 @@ func (dhtPeer *DHTPeer) handleDelegateService(ready *sync.WaitGroup) {
...
 
+	done := false
 	for {
 		select {
 		default:
 			linfo().Msg("DelegateService listening for new connections...")
-			ready.Done()
+			if !done {
+				done = true
+				ready.Done()
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:348]: Notification mechanism to guarantee  that delegation service is running before `New` returns the newly created `DHTPeer`

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go [hunk 2]
```diff
@@ -671,12 +675,28 @@ func (dhtPeer *DHTPeer) handleAeaAddressStream(stream network.Stream) {
...
 
 func (dhtPeer *DHTPeer) handleAeaNotifStream(stream network.Stream) {
-	lerror, _, linfo, _ := dhtPeer.getLoggers()
+	lerror, _, linfo, ldebug := dhtPeer.getLoggers()
 
 	linfo().Str("op", "notif").
 		Msgf("Got a new notif stream")
 
 	if !dhtPeer.addressAnnounced {
+		// workaround: to avoid getting `failed to find any peer in table`
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:684]: Fixes race condition that causes libp2p `test_communication.py` to be flaky.
It seems that the race condition is well know to libp2p [1].
They even need to consider it in their own tests [2]

[1] https://github.com/libp2p/go-libp2p-kad-dht/issues/308
[2] https://github.com/libp2p/go-libp2p-kad-dht/blob/v0.7.11/dht_test.go#L179

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go [hunk 3]
```diff
@@ -67,38 +73,12 @@ var (
...
-	select {
-	case envel := <-rx:
-		t.Log("Received envelope", envel)
-	case <-timeout:
-		t.Error("Failed to receive envelope before timeout")
-	}
-}
+/*
+	DHT Network: DHTPeer-to-DHTPeer
+*/
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go:78]: Tests different patterns of `DHTPeer` to `DHTPeer` communications

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go [hunk 4]
```diff
@@ -279,22 +330,1073 @@ func TestRoutingStarFullConnectivity(t *testing.T) {
...
-			expectEnvelope(t, rxs[j])
-			expectEnvelope(t, rxs[i])
+			expectEnvelope(t, rxs[target])
+			expectEnvelope(t, rxs[from])
 		}
 	}
 }
+
+/*
+	DHT network: DHTClient
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go:358]: Tests different patterns of `DHTClient` to {`DHTPeer`, `DHTClient`} communications

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go [hunk 5]
```diff
@@ -279,22 +330,1073 @@ func TestRoutingStarFullConnectivity(t *testing.T) {
...
+		t.Error("Failed to RouteEnvelope from DHTClient to DHTClient:", err)
+	}
+
+	expectEnvelope(t, rxClient1)
+	expectEnvelope(t, rxClient2)
+
+}
+
+/*
+	DHT network: DelegateClient
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go:588]: Tests different patterns of `DelegateClient` to {`DHTPeer`, `DHTClient`, `DelegateClient`} communications

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go [hunk 6]
```diff
@@ -279,22 +330,1073 @@ func TestRoutingStarFullConnectivity(t *testing.T) {
...
+
+   DHTClient -------                                                 -- DelegateClient
+                   |                                                 |
+   DHTClient -------                                                 -- DelegateClient
+                   |                                                 |
+                   |-- DHTPeer --- DHTPeeer -- DHTPeer --- DHTPeer --|
+                   |                                                 |
+   DelegateClient --                                                 ------- DHTClient
+*/
+
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer_test.go:971]: Tests all possible communication patterns at once

File: packages/fetchai/connections/p2p_libp2p/libp2p_node.go [hunk 7]
```diff
@@ -21,43 +21,22 @@
 package main
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/libp2p_node.go:21]: Deprecate old code

File: packages/fetchai/connections/p2p_libp2p/utils/utils.go [hunk 8]
```diff
@@ -105,6 +135,22 @@ func BootstrapConnect(ctx context.Context, ph host.Host, peers []peer.AddrInfo)
 	if count == len(peers) {
 		return errors.New("failed to bootstrap: " + err.Error())
 	}
+
+	// workaround: to avoid getting `failed to find any peer in table`
+	//  when calling dht.Provide (happens occasionally)
+	logger.Debug().Msg("waiting for bootstrap peers to be added to dht routing table...")
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/utils/utils.go:141]: See above

File: tests/test_cli/test_generate/test_protocols.py [hunk 9]
```diff
@@ -45,7 +45,7 @@
 )
```
Comment (lrahmani) [inline tests/test_cli/test_generate/test_protocols.py:45]: Disabled due to consistently failing.

@DavidMinarsch sould be re-enabled before merging? or keep disabled until fixed?

File: tests/test_crypto/test_cosmos_crypto.py [hunk 10]
```diff
@@ -80,7 +80,7 @@ def test_api_none():
```
Comment (lrahmani) [inline tests/test_crypto/test_cosmos_crypto.py:80]: Disabled due to consistently failing.

Sould be re-enabled before merging

File: tests/test_packages/test_connections/test_p2p_libp2p/test_communication.py [hunk 11]
```diff
@@ -31,7 +31,6 @@
 from aea.protocols.default.message import DefaultMessage
 
 from ....conftest import (
-    MAX_FLAKY_RERUNS,
```
Comment (lrahmani) [inline tests/test_packages/test_connections/test_p2p_libp2p/test_communication.py:34]: Remove flaky mark, as communication tests should be 100% reliable with fix included in this pr

File: tests/test_packages/test_connections/test_p2p_libp2p_client/test_communication.py [hunk 12]
```diff
@@ -31,7 +31,6 @@
 from aea.protocols.default.serialization import DefaultSerializer
 
 from ....conftest import (
-    MAX_FLAKY_RERUNS,
```
Comment (lrahmani) [inline tests/test_packages/test_connections/test_p2p_libp2p_client/test_communication.py:34]: Remove flaky mark, as communication tests should be 100% reliable with fix included in this pr

File: .github/workflows/workflow.yml [hunk 13]
```diff
@@ -209,5 +210,28 @@ jobs:
     - if: matrix.python-version == '3.6' && matrix.os != 'windows-latest'
       name: Golang unit tests
       working-directory: ./packages/fetchai/connections/p2p_libp2p
-      run: go test -p 1 -count 2 -v ./...
+      run: go test -p 1 -timeout 0 -count 1 -v ./...
         
+  golang_checks:
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply .github/workflows/workflow.yml:215]: Why don't we just run all go tests here and remove them from platform checks? Then we have more of a separation.

File: docs/p2p-connection.md [hunk 14]
```diff
@@ -123,20 +123,19 @@ aea run --connections "fetchai/p2p_libp2p:0.2.0,fetchai/oef:0.5.0"
...
 
-In particular, by modiying the configuration such that:
+In particular, by modifying the configuration such that:
 ``` yaml
 config:
   delegate_uri: 127.0.0.1:11001
   entry_peers: [/dns4/agents-p2p-dht.sandbox.fetch-ai.com/tcp/9000/p2p/16Uiu2HAkw1ypeQYQbRFV5hKUxGRHocwU5ohmVmCnyJNg36tnPFdx, /dns4/agents-p2p-dht.sandbox.fetch-ai.com/tcp/9001/p2p/16Uiu2HAmVWnopQAqq4pniYLw44VRvYxBUoRHqjz1Hh2SoCyjbyRW, /dns4/agents-p2p-dht.sandbox.fetch-ai.com/tcp/9002/p2p/16Uiu2HAmNJ8ZPRaXgYjhFf8xo8RBTX8YoUU5kzTW7Z4E5J3x9L1t]
   local_uri: 127.0.0.1:9001
   log_file: libp2p_node.log
-  public_uri: 127.0.0.1:9001
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply docs/p2p-connection.md:141]: So probably we should implement a code level check for this on the python side?

File: packages/fetchai/connections/p2p_libp2p/aea/api.go [hunk 15]
```diff
@@ -420,97 +437,3 @@ func run_aea_sandbox(msgin_path string, msgout_path string) error {
 	return nil
 }
 */
-
-/*
-
-  Protobuf generated Envelope - Edited
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply packages/fetchai/connections/p2p_libp2p/aea/api.go:426]: Nice!!!

File: packages/fetchai/connections/p2p_libp2p/aea/envelope.proto [hunk 16]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply packages/fetchai/connections/p2p_libp2p/aea/envelope.proto:1]: OK

File: packages/fetchai/connections/p2p_libp2p/aea/envelope.proto [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+package aea;
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/aea/envelope.proto:3]: We need to come up with a proper package naming convention

Comment (DavidMinarsch) [reply packages/fetchai/connections/p2p_libp2p/aea/envelope.proto:3]: Also relates to protocols we generate for AEA to AEA comms

File: packages/fetchai/connections/p2p_libp2p/go.sum [hunk 18]
```diff
@@ -20,6 +20,7 @@ github.com/btcsuite/goleveldb v0.0.0-20160330041536-7834afc9e8cd/go.mod h1:F+uVa
 github.com/btcsuite/snappy-go v0.0.0-20151229074030-0bdef8d06723/go.mod h1:8woku9dyThutzjeg+3xrA5iCpBRH8XEEg3lh6TiUghc=
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/go.sum:20]: @lrahmani please check if there is a way to automatically check for licenses in the dependencies. We do this on the python side and it is quite important obviously.

File: tests/test_cli/test_generate/test_protocols.py (hunk 9 above)
Comment (DavidMinarsch) [reply tests/test_cli/test_generate/test_protocols.py:45]: Reenable please;

File: tests/test_crypto/test_cosmos_crypto.py (hunk 10 above)
Comment (DavidMinarsch) [reply tests/test_crypto/test_cosmos_crypto.py:80]: Please enable; this is just an account funding issue; you can fund the account yourself when it happens again (it is now funded) by finding the private key and generating some test wealth for it.

File: .github/workflows/workflow.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (lrahmani) [reply .github/workflows/workflow.yml:215]: will do that

File: docs/p2p-connection.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (lrahmani) [reply docs/p2p-connection.md:141]: I have been thinking of it, still didn't settle down on the approach
Basically, checking it translates to making sure that the `public_uri` and `entry_peers` dns-es can be either all (i) private[1] ip addresses or all (ii) public ip addresses

[1] https://en.wikipedia.org/wiki/Private_network#Private_IPv4_addresses

File: packages/fetchai/connections/p2p_libp2p/aea/envelope.proto (hunk 17 above)
Comment (lrahmani) [reply packages/fetchai/connections/p2p_libp2p/aea/envelope.proto:3]: Agree

File: packages/fetchai/connections/p2p_libp2p/go.sum (hunk 18 above)
Comment (lrahmani) [reply packages/fetchai/connections/p2p_libp2p/go.sum:20]: good remark, will look into it

**Pull request merged**